algorithms/src/std_algorithms/numeric/Kokkos_Reduce.hpp

```diff
@@ -142,10 +140,24 @@ ValueType reduce_default_functors_impl(const std::string& label,
   Impl::static_assert_is_not_openmptarget(ex);
   Impl::expect_valid_range(first, last);
 
-  using value_type  = Kokkos::Impl::remove_cvref_t<ValueType>;
-  using joiner_type = Impl::StdReduceDefaultJoinFunctor<value_type>;
-  return reduce_custom_functors_impl(
-      label, ex, first, last, std::move(init_reduction_value), joiner_type());
+  if (first == last) {
+    // init is returned, unmodified
+    return init_reduction_value;
+  }
+
+  using value_type   = Kokkos::Impl::remove_cvref_t<ValueType>;
+  using functor_type = Impl::StdReduceDefaultFunctor<IteratorType, value_type>;
+
+  // run
+  value_type tmp;
+  const auto num_elements = Kokkos::Experimental::distance(first, last);
+  ::Kokkos::parallel_reduce(label,
+                            RangePolicy<ExecutionSpace>(ex, 0, num_elements),
+                            functor_type{first}, tmp);
+
+  // fence not needed since reducing into scalar
+  tmp += init_reduction_value;
+  return tmp;
 }
```

Comment:
i think that if we did this, we would have the same problem as in the inclusive_scan PR that it won't work for a value type that does not have a reduction identity.